logging in via webview doesn't work

when trying to login with webview the launcher crashes tried with both java 8 & 17:
```
java.lang.NoClassDefFoundError: javafx/embed/swing/JFXPanel
	at fr.litarvan.openauth.microsoft.MicrosoftAuthenticator.loginWithAsyncWebview(MicrosoftAuthenticator.java:151)
	at fr.litarvan.openauth.microsoft.MicrosoftAuthenticator.loginWithWebview(MicrosoftAuthenticator.java:138)
	at me.earth.headlessmc.launcher.auth.AccountManager.loginWithWebView(AccountManager.java:69)
	at me.earth.headlessmc.launcher.command.login.LoginCommand.loginWithWebView(LoginCommand.java:75)
	at me.earth.headlessmc.launcher.command.login.LoginCommand.execute(LoginCommand.java:25)
	at me.earth.headlessmc.command.CommandContextImpl.executeCommand(CommandContextImpl.java:44)
	at me.earth.headlessmc.command.CommandContextImpl.execute(CommandContextImpl.java:31)
	at me.earth.headlessmc.command.line.ConsoleListener.listen(ConsoleListener.java:22)
	at me.earth.headlessmc.command.line.CommandLineImpl.listen(CommandLineImpl.java:26)
	at me.earth.headlessmc.launcher.Main.runHeadlessMc(Main.java:90)
	at me.earth.headlessmc.launcher.Main.main(Main.java:35)
Caused by: java.lang.ClassNotFoundException: javafx.embed.swing.JFXPanel
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	... 11 more
```

```
$ java -version
openjdk version "17.0.9" 2023-10-17
OpenJDK Runtime Environment (build 17.0.9+8)
OpenJDK 64-Bit Server VM (build 17.0.9+8, mixed mode)
```
```
$ java -version
openjdk version "1.8.0_392"
OpenJDK Runtime Environment (build 1.8.0_392-b08)
OpenJDK 64-Bit Server VM (build 25.392-b08, mixed mode)
```
```
$ uname -nr
archlinux 6.5.8-arch1-1
```

I got it to work in dev environment by adding the javafx plugin and using it to add the missing libs but couldn't get it working when building a jar.

--- a/.gitignore
+++ b/.gitignore
@@ -24,6 +24,7 @@ run
 # HeadlessMC
 /HeadlessMC/
 /headlessmc-launcher/HeadlessMC/
+/headlessmc-launcher-jfx/HeadlessMC/
 
 # Created by some launcher tests
 /headlessmc-launcher/test/

--- a/build.gradle
+++ b/build.gradle
@@ -157,6 +157,13 @@ allprojects {
                     ((MavenPublication) it).artifactId "${project.archivesBaseName.toLowerCase()}"
                     ((MavenPublication) it).version "${project.version}${pubSuffix}"
                     // TODO: testFixtures are part of commons artifact?
+                    // Exclude shadowJars from publication
+                    components.withType(AdhocComponentWithVariants).forEach { c ->
+                        c.withVariantsFromConfiguration(project.configurations.shadowRuntimeElements) {
+                            skip()
+                        }
+                    }
+
                     from components.java
                 }
             }

--- a/build.gradle
+++ b/build.gradle
@@ -184,6 +184,8 @@ allprojects {
         publish {
             dependsOn(build)
         }
+    } else {
+        project.logger.warn('Cannot publish on Java 9+ as we are not generating JavaDocs!')
     }
 }
 

--- a/build.gradle
+++ b/build.gradle
@@ -213,7 +213,13 @@ tasks.register('jacocoRootReport', JacocoReport) {
 
 tasks.register('copyJars', Copy) {
     dependsOn subprojects.shadowJar
-    from(subprojects.shadowJar)
+    subprojects.each { subproject ->
+        // jfx jar is > 100MB, I do not want to fill my github storage?
+        if (subproject.name != 'headlessmc-launcher-jfx') {
+            from(subproject.shadowJar)
+        }
+    }
+
     into project.file(System.getProperty('hmc.jar.dir', 'build/libs'))
 }
 

--- a/buildSrc/src/main/groovy/me/earth/headlessmc/gradle/Extension2ClassWriterAdapter.groovy
+++ b/buildSrc/src/main/groovy/me/earth/headlessmc/gradle/Extension2ClassWriterAdapter.groovy
@@ -1,10 +1,10 @@
 package me.earth.headlessmc.gradle
 
+import org.objectweb.asm.ClassWriter
+
 import static me.earth.headlessmc.gradle.ModuleExtension.ACC_MANDATED
 import static org.objectweb.asm.Opcodes.ACC_MODULE
 import static org.objectweb.asm.Opcodes.V9
-
-import org.objectweb.asm.ClassWriter
 
 class Extension2ClassWriterAdapter {
     static ClassWriter toClassWriter(ModuleExtension me) {

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,1 +1,1 @@
-project_version=1.7.3
+project_version=1.8.0

--- a/headlessmc-commons/src/main/java/me/earth/headlessmc/logging/LoggingHandler.java
+++ b/headlessmc-commons/src/main/java/me/earth/headlessmc/logging/LoggingHandler.java
@@ -5,10 +5,9 @@ import me.earth.headlessmc.config.HmcProperties;
 import me.earth.headlessmc.util.ResourceUtil;
 
 import java.io.*;
-import java.util.logging.Level;
-import java.util.logging.LogManager;
-import java.util.logging.LogRecord;
-import java.util.logging.StreamHandler;
+import java.util.Iterator;
+import java.util.logging.*;
+import java.util.logging.Logger;
 
 /**
  * A {@link StreamHandler} logging to {@link FileDescriptor#out} using a {@link
@@ -21,9 +20,23 @@ public class LoggingHandler extends StreamHandler {
     }
 
     public static void apply() throws IOException {
-        @Cleanup
-        InputStream is = ResourceUtil.getHmcResource("logging.properties");
-        LogManager.getLogManager().readConfiguration(is);
+        try {
+            // check LoggingHandler can actually load the class from the SystemClassLoader
+            ClassLoader.getSystemClassLoader().loadClass(LoggingHandler.class.getName());
+            @Cleanup
+            InputStream is = ResourceUtil.getHmcResource("logging.properties");
+            LogManager.getLogManager().readConfiguration(is);
+        } catch (Exception ignored) {
+            Handler[] handlers = Logger.getLogger("").getHandlers();
+            for (Handler handler : handlers) {
+                if (handler != null && !(handler instanceof FileHandler)) {
+                    Logger.getLogger("").removeHandler(handler);
+                }
+            }
+            // add the Handler manually
+            Logger.getLogger("").addHandler(new LoggingHandler());
+        }
+
         String property = System.getProperty(HmcProperties.LOGLEVEL.getName());
         if (property == null || !LogLevelUtil.trySetLevel(property)) {
             LogLevelUtil.setLevel(Level.INFO);

--- a/headlessmc-commons/src/main/java/me/earth/headlessmc/logging/LoggingHandler.java
+++ b/headlessmc-commons/src/main/java/me/earth/headlessmc/logging/LoggingHandler.java
@@ -5,9 +5,8 @@ import me.earth.headlessmc.config.HmcProperties;
 import me.earth.headlessmc.util.ResourceUtil;
 
 import java.io.*;
-import java.util.Iterator;
-import java.util.logging.*;
 import java.util.logging.Logger;
+import java.util.logging.*;
 
 /**
  * A {@link StreamHandler} logging to {@link FileDescriptor#out} using a {@link

--- a/headlessmc-commons/src/test/java/me/earth/headlessmc/command/CommandUtilTest.java
+++ b/headlessmc-commons/src/test/java/me/earth/headlessmc/command/CommandUtilTest.java
@@ -1,7 +1,6 @@
 package me.earth.headlessmc.command;
 
 import me.earth.headlessmc.util.AbstractUtilityTest;
-import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 import java.util.HashMap;

--- a/headlessmc-launcher-jfx/build.gradle
+++ b/headlessmc-launcher-jfx/build.gradle
@@ -1,0 +1,60 @@
+plugins {
+    id 'application'
+    id 'headlessmc-modules'
+}
+
+repositories {
+    maven {
+        name 'OpenAuthMaven'
+        url 'https://litarvan.github.io/maven'
+    }
+}
+
+def MAIN_CLASS = 'me.earth.headlessmc.launcher.Main'
+application {
+    mainClass = MAIN_CLASS
+}
+
+run {
+    standardInput = System.in
+}
+
+module {
+    name = 'headlessmc.launcher'
+    access = ACC_OPEN
+    // TODO: otherwise we include the gson module-info, which is also not nice
+}
+
+def javaFxVersion = '17.0.9'
+version += '-' + javaFxVersion
+
+dependencies {
+    jarLibsApi project(':headlessmc-launcher')
+    jarLibs group: 'org.openjfx', name: 'javafx-base', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-base', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-base', version: javaFxVersion, classifier: 'linux'
+    jarLibs group: 'org.openjfx', name: 'javafx-controls', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-controls', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-controls', version: javaFxVersion, classifier: 'linux'
+    jarLibs group: 'org.openjfx', name: 'javafx-fxml', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-fxml', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-fxml', version: javaFxVersion, classifier: 'linux'
+    jarLibs group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'linux'
+    jarLibs group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'linux'
+    jarLibs group: 'org.openjfx', name: 'javafx-web', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-web', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-web', version: javaFxVersion, classifier: 'linux'
+    jarLibs group: 'org.openjfx', name: 'javafx-media', version: javaFxVersion, classifier: 'win'
+    jarLibs group: 'org.openjfx', name: 'javafx-media', version: javaFxVersion, classifier: 'mac'
+    jarLibs group: 'org.openjfx', name: 'javafx-media', version: javaFxVersion, classifier: 'linux'
+}
+
+shadowJar {
+    excludes.remove("module-info.class")
+    exclude("META-INF/maven/com.google.code.gson/**")
+    exclude("LICENSE")
+}

--- a/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/Launcher.java
+++ b/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/Launcher.java
@@ -15,7 +15,7 @@ import me.earth.headlessmc.launcher.version.VersionService;
 @Getter
 @RequiredArgsConstructor
 public class Launcher implements HeadlessMc {
-    public static final String VERSION = "1.7.3";
+    public static final String VERSION = "1.8.0";
 
     @Delegate
     private final HeadlessMc headlessMc;

--- a/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/auth/AccountManager.java
+++ b/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/auth/AccountManager.java
@@ -67,6 +67,10 @@ public class AccountManager implements Iterable<Account> {
         val authenticator = new MicrosoftAuthenticator();
         try {
             val result = authenticator.loginWithWebview();
+            if (result == null) {
+                throw new AuthException("Did not login into any account.");
+            }
+
             return setAccount(result);
         } catch (MicrosoftAuthenticationException e) {
             throw new AuthException(e.getMessage());

--- a/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/command/download/VersionInfoCache.java
+++ b/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/command/download/VersionInfoCache.java
@@ -4,7 +4,6 @@ import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import lombok.CustomLog;
 import lombok.Getter;
-import lombok.SneakyThrows;
 import lombok.experimental.Delegate;
 import lombok.val;
 import me.earth.headlessmc.launcher.util.JsonUtil;

--- a/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/command/login/LoginCommand.java
+++ b/headlessmc-launcher/src/main/java/me/earth/headlessmc/launcher/command/login/LoginCommand.java
@@ -77,6 +77,9 @@ public class LoginCommand extends AbstractLauncherCommand {
                         + " successfully.");
         } catch (AuthException e) {
             ctx.log("Failed to log in: " + e.getMessage());
+        } catch (NoClassDefFoundError e) {
+            log.debug(e.getMessage());
+            ctx.log("Your version of Java does not support Webview! It usually comes bundled with JDK 8 or in the headlessmc-launcher-jfx jar.");
         }
     }
 

--- a/headlessmc-launcher/src/test/java/me/earth/headlessmc/launcher/launch/ProcessFactoryTest.java
+++ b/headlessmc-launcher/src/test/java/me/earth/headlessmc/launcher/launch/ProcessFactoryTest.java
@@ -7,7 +7,9 @@ import me.earth.headlessmc.launcher.LauncherMock;
 import me.earth.headlessmc.launcher.os.OS;
 import org.junit.jupiter.api.Test;
 
-import java.io.*;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 

--- a/headlessmc-scripts/hmc
+++ b/headlessmc-scripts/hmc
@@ -1,2 +1,2 @@
 #!/usr/bin/env sh
-java -jar headlessmc-launcher-1.7.3.jar --command $@
+java -jar headlessmc-launcher-1.8.0.jar --command $@

--- a/headlessmc-scripts/hmc.bat
+++ b/headlessmc-scripts/hmc.bat
@@ -1,2 +1,2 @@
 @echo off
-"%JAVA_HOME%\bin\java" -jar headlessmc-launcher-1.7.3.jar --command %*
+"%JAVA_HOME%\bin\java" -jar headlessmc-launcher-1.8.0.jar --command %*

--- a/headlessmc-scripts/hmw
+++ b/headlessmc-scripts/hmw
@@ -1,3 +1,3 @@
 #!/bin/bash
 # when running in docker on windows bash seems to be at /bin/bash TODO: can we make this one script?
-java -jar headlessmc-launcher-1.7.3.jar --command $@
+java -jar headlessmc-launcher-1.8.0.jar --command $@

--- a/settings.gradle
+++ b/settings.gradle
@@ -4,4 +4,4 @@ include 'headlessmc-commons'
 include 'headlessmc-runtime'
 include 'headlessmc-lwjgl'
 include 'headlessmc-launcher'
-
+include 'headlessmc-launcher-jfx'

